pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/RuntimeASTTransformer.groovy

```diff
@@ -1238,10 +1249,37 @@ class RuntimeASTTransformer {
             // These variable handles are declared in functions, but will still be bound to the script context
             // Doing this here ensures the variables make the trip across to run time - if they don't make it, neither did this pipeline
             pipelineBlock.addStatement(stmt(callX(closureX(currentBlock), 'call')))
-            pipelineElementHandles.clear()
         }
 
-        /**
+        @NonNull
+        private List<Statement> prepareClosureScopedHandles(@NonNull BlockStatement pipelineBlock) {
+            ArrayList<Statement> result = new ArrayList<Statement>()
+            if (SCRIPT_SPLITTING_TRANSFORMATION) {
+                ArrayList<DeclarationExpression> declarations = new ArrayList<DeclarationExpression>()
+                moduleNode.statementBlock.statements.each { item ->
+                    if (((ExpressionStatement) item).expression instanceof DeclarationExpression) {
+                        declarations.add((DeclarationExpression) ((ExpressionStatement) item).expression)
+                    }
+                }
+
+                if (declarations.size() != 0) {
+                    if (SCRIPT_SPLITTING_ALLOW_LOCAL_VARIABLES) {
+                        result.addAll(pipelineElementHandles)
+                        pipelineElementHandles.clear()
+                    } else {
+                        throw new IllegalStateException("[JENKINS-34987] SCRIPT_SPLITTING_TRANSFORMATION is an experimental feature of Declarative Pipeline and is incompatible with local variable declarations inside a Jenkinsfile. " +
+                                "Add the '@Field' annotation to local variable declarations: " +
```

Comment:
This recommendation seems weird to me. Even if turning local variables into fields using `@Field` makes it possible to do the script splitting transformation, I don't think it's something we should recommend, since this message makes it sound like using `@Field` is something we endorse as a good idea when really we do not recommend using Groovy variables with Declarative at all.

Reply:
However, we do not provide any alternative facility, so this is what people do.  Given the number of people that have continued to complain about this issue, it is clear that there are plenty of people using Declarative with variables (otherwise the existing feature would probably have worked for them).  I'll update the message to be clear that this is not supported.